src/high-level/matrix.lisp

```diff
@@ -0,0 +1,536 @@
+                               ,"The shape of ~a is ~a, which is not a square" ,(symbol-name matrix) (shape ,matrix)))))
+
+;;; Required abstract-tensor methods
+
+(defmethod rank ((m matrix))
+  (declare (ignore m))
+  2)
+
+(defmethod shape ((m matrix))
+  (list (matrix-nrows m) (matrix-ncols m)))
+
+(defmethod (setf shape) (new-value (m matrix))
+  (policy-cond:policy-if
+   (< speed safety)
+   (progn
+     (check-type new-value shape)
+     (assert (cl:= 2 (length new-value))
+             () "Matrix shape must be of length 2"))
+   nil)
+  (setf (matrix-nrows m) (first new-value)
+        (matrix-ncols m) (second new-value)))
+
+;; Specific constructors
+
+(defgeneric random-unitary (shape &key type)
+  (:documentation "Generate a uniformly random element of U(n).")
+  (:method (shape &key (type +default-tensor-type+))
+    (policy-cond:policy-if
+     (< speed safety)
+     (assert-square-shape shape)
+     nil)
+    (multiple-value-bind (q r) (qr (rand shape :type type :distribution #'alexandria:gaussian-random))
+      (let ((d (diag r)))
+        (setf d (cl:map 'list (lambda (di) (/ di (sqrt (* di (conjugate di))))) d))
+        (@ q (funcall #'from-diag d shape))))))
+
+;; TODO: This should be generic to abstract-tensor
+(defgeneric ptr-ref (m base i j)
+  (:documentation
+   "Accessor method for the pointer to the element in the I-th row and J-th column of a matrix M, assuming zero indexing.")
+  (:method ((m matrix) base i j)
+    (policy-cond:policy-if
+     (< speed safety)
+     (assert (valid-index-p (list i j) (shape m))
+             () "Incompatible position for MATRIX. Position ~a is not within matrix shape ~a" (list i j) (shape m))
+     nil)
+    (let ((type (element-type m)))
+      ;; TODO: compensate for order
+      (let ((idx (column-major-index (list i j) (shape m))))
+        (cond
+          ((subtypep type 'single-float) (cffi:mem-aptr base :float idx))
+          ((subtypep type 'double-float) (cffi:mem-aptr base :double idx))
+          ((subtypep type '(complex single-float)) (cffi:mem-aptr base :float (* 2 idx)))
+          ((subtypep type '(complex double-float)) (cffi:mem-aptr base :double (* 2 idx)))
+          (t (error "Incompatible element type ~a." type)))))))
+
+(defgeneric row (matrix index)
+  (:documentation "Get row vector from a matrix")
+  (:method ((m matrix) index)
+    (check-type index alexandria:non-negative-fixnum)
+    (slice m
+           (list index 0)
+           (list (1+ index) (ncols m)))))
+
+(defgeneric column (matrix index)
+  (:documentation "Get column vector from a matrix")
+  (:method ((m matrix) index)
+    (slice m
+           (list 0 index)
+           (list (nrows m) (1+ index)))))
+
+(defgeneric mult (a b &key target alpha beta transa transb)
+  (:documentation "Multiply matrix a by matrix b, storing in target or creating a new matrix if target is not specified.
+Target cannot be the same as a or b."))
+
+(defgeneric @ (matrix &rest matrices)
```

Comment:
For the wish list: generic support for matrix-vector multiplication, e.g.
```
(magicl:@ (magicl:from-list '(0d0 1d0 2d0 3d0) '(2 2)) 
                  (magicl:from-list '(4d0 5d0) '(2)))
```
which currently gives an error, but works if instead the rightmost `VECTOR` is instead a `MATRIX` with shape `(2 1).

Reply:
As a related question, what is the appropriate way for a user of this API to reshape a vector to a matrix (as a 1xN or Nx1 matrix, lets sa)?

Reply:
`reshape` would be the best one, maybe with `generalize-tensor`. I am looking into methods of broadcasting for this. `reshape` and `generalize-tensor` were also broken with the shift from clos to structs

Reply:
It should work now with `(magicl:reshape (magicl:from-list '(4d0 5d0) '(2)) '(2 1))`

Reply:
#82